File: 1655_가운데를_말해요.py
'''
left - 최대힙 (음수로 넣음)
right - 최소힙
left와 right 길이가 같으면 left로 넣기
개수가 짝수일 경우 작은 값을 꺼내야 하므로 힙에 있는 값 비교해서 바꿔주기
답은 left에서 꺼내면 된다.
'''
import heapq
import sys

n = int(sys.stdin.readline())
left = []
right = []

def printMiddleNumber(x):
    if len(left) == len(right):
        heapq.heappush(left, -x)
    else:
        heapq.heappush(right,x)
    if left and right and -left[0] > right[0]:
        left_temp = heapq.heappop(left) * -1
        right_temp = heapq.heappop(right)

        heapq.heappush(left, -right_temp)
        heapq.heappush(right, left_temp)
    print(-left[0])
for _ in range(n):
    printMiddleNumber(int(sys.stdin.readline()))

File: 1655_가운데를_말해요.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        PriorityQueue<Integer> left = new PriorityQueue<>();     // 최대힙
        PriorityQueue<Integer> right = new PriorityQueue<>();    // 최소힙

        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            int x = Integer.parseInt(br.readLine());

            if (left.size() == right.size()) {
                left.add(-x);
            } else {
                right.add(x);
            }

            if (!left.isEmpty() && !right.isEmpty() && -left.peek() > right.peek()) {
                int left_temp = -left.poll();
                int right_temp = right.poll();

                left.add(-right_temp);
                right.add(left_temp);
            }

            sb.append(-left.peek()).append("\n");
        }
        System.out.println(sb);
    }
}